plugins/renepay/flow.c

```diff
@@ -0,0 +1,849 @@
+	{
+		debug_err("%s (line %d) B=%s must be greater than A=%s",
+			__PRETTY_FUNCTION__,
+			__LINE__,
+			type_to_string(this_ctx, struct amount_msat, &B),
+			type_to_string(this_ctx, struct amount_msat, &A));
+	}
+	struct amount_msat numerator; // MAX(0,B-f)
+
+	if(!amount_msat_sub(&numerator,B,f))
+		numerator = AMOUNT_MSAT(0);
+
+	tal_free(this_ctx);
+	return amount_msat_less_eq(f,A) ? 1.0 : amount_msat_ratio(numerator,denominator);
+}
+
+
+
+
+
+
+// TODO(eduardo): remove this function, is a duplicate
+void remove_completed_flow(const struct gossmap *gossmap,
+			   struct chan_extra_map *chan_extra_map,
+			   struct flow *flow)
+{
+	for (size_t i = 0; i < tal_count(flow->path); i++) {
+		struct chan_extra_half *h = get_chan_extra_half_by_chan(gossmap,
+							       chan_extra_map,
+							       flow->path[i],
+							       flow->dirs[i]);
+		if (!amount_msat_sub(&h->htlc_total, h->htlc_total, flow->amounts[i]))
+		{
+			debug_err("%s could not substract HTLC amounts, "
+				   "half total htlc amount = %s, "
+				   "flow->amounts[%lld] = %s.",
+				   __PRETTY_FUNCTION__,
+				   type_to_string(tmpctx, struct amount_msat, &h->htlc_total),
+				   i,
+				   type_to_string(tmpctx, struct amount_msat, &flow->amounts[i]));
+		}
+		if (h->num_htlcs == 0)
+		{
+			debug_err("%s could not decrease HTLC count.",
+				   __PRETTY_FUNCTION__);
+		}
+		h->num_htlcs--;
+	}
+}
+// TODO(eduardo): remove this function, is a duplicate
+void remove_completed_flow_set(
+		const struct gossmap *gossmap,
+		struct chan_extra_map *chan_extra_map,
+		struct flow **flows)
+{
+	for(size_t i=0;i<tal_count(flows);++i)
+	{
+		remove_completed_flow(gossmap,chan_extra_map,flows[i]);
+	}
+}
+
+// TODO(eduardo): remove this function, is a duplicate
+void commit_flow(
+		const struct gossmap *gossmap,
+		struct chan_extra_map *chan_extra_map,
+		struct flow *flow)
+{
+	for (size_t i = 0; i < tal_count(flow->path); i++) {
+		struct chan_extra_half *h = get_chan_extra_half_by_chan(gossmap,
+							       chan_extra_map,
+							       flow->path[i],
+							       flow->dirs[i]);
+		if (!amount_msat_add(&h->htlc_total, h->htlc_total, flow->amounts[i]))
+		{
+			debug_err("%s could not add HTLC amounts, "
+				   "flow->amounts[%lld] = %s.",
+				   __PRETTY_FUNCTION__,
+				   i,
+				   type_to_string(tmpctx, struct amount_msat, &flow->amounts[i]));
+		}
+		h->num_htlcs++;
+	}
+}
+// TODO(eduardo): remove this function, is a duplicate
+void commit_flow_set(
+		const struct gossmap *gossmap,
+		struct chan_extra_map *chan_extra_map,
+		struct flow **flows)
+{
+	for(size_t i=0;i<tal_count(flows);++i)
+	{
+		commit_flow(gossmap,chan_extra_map,flows[i]);
+	}
+}
+
+/* Helper function to fill in amounts and success_prob for flow
+ *
+ * IMPORTANT: here we do not commit flows to chan_extra, flows are commited
+ * after we send those htlc.
+ *
+ * IMPORTANT: flow->success_prob is misleading, because that's the prob. of
```

Comment:
the python code that I wrote has a global state of allocated HTLCs. Of course that does not respect HTLCs allocated from other participants in the network. Still why counting allocated HTLCs only for one active payment(session) and not together with the other ongoing payments?

Reply:
We should take into account the known allocated liquidity in HTLCs, I think that's crucial! Thanks for pointing this.
I'll fix it.

Reply:
Fixed https://github.com/ElementsProject/lightning/pull/6376/commits/b0ca3f1b90654d290b3304990454b537110d7738